### ISSUE TITLE
[docs] Code in "Asynchronous Data Queries" has a bug

### ISSUE THREAD

**Issue by acutmore** [facebookexperimental/Recoil#191]
The code shown in https://recoiljs.org/docs/guides/asynchronous-data-queries/ doesn't work.

A selector is returning a promise, but then being used synchronously in the next selector.

```javascript
const userInfoQuery = selectorFamily({
  key: "UserInfoQuery",
  get: (userID) => async ({ get }) => {
    const response = await myDBQuery({ userID });
    if (response.error) {
      throw response.error;
    }
    return response;
  },
});

const currentUserInfoQuery = selector({
  key: "CurrentUserInfoQuery",
  // userInfoQuery returns a promise, so currentUserInfoQuery also returns a promise
  get: ({ get }) => get(userInfoQuery(get(currentUserIDState)), // missing bracket here
});

const friendsInfoQuery = selector({
  key: "FriendsInfoQuery",
  get: ({ get }) => {
    // currentUserInfoQuery returns a promise, so we need to await the result here
    const { friendList } = get(currentUserInfoQuery);
    const friends = [];
    for (const friendID of friendList) {
      const friendInfo = get(userInfoQuery(friendID));
      friends.push(friendInfo);
    }
    return friends; // this will be an array or promises
  },
});
```

When I first read this page in the docs I actually really liked the code. I assumed Recoil was doing something similar to React, where accessing a selector that returns a Promise would throw, and then _restart_ the selector when the promise resolves. But when trying to run, it looks like if a selector is changed to return a promise then all selectors downstream need to be updated to await the result, which is a little bit of a shame.

**Comment by drarmstr** [comment]
A synchronous selector can use an asynchronous selector as a dependency.  It is part of the design that you can read the value from any Recoil state and the interface is the same: the implementation can change between being an atom, selector, or asynchronous selector without changing code using the state.

If a synchronous selector attempts to use an asynchronous selector that is still pending or throws an error then it will propagate that loading or error state.  It is not necessary to make the downstream selectors "async".  Under the covers, when it tries to get a pending selector it will throw and then restart when the dependency resolves or errors.

There is a known issue that dependencies observed after the first async dependency are not recorded.  They will work, but won't subscribe to update.  See #103  We have a fix for this (See "selector_NEW") But, there are some issues and we need to fix up some CM support first before we can deploy it.  That's in progress.

**Comment by acutmore** [comment]
> If a synchronous selector attempts to use an asynchronous selector that is still pending or throws an error then it will propagate that loading or error state. It is not necessary to make the downstream selectors "async". Under the covers, when it tries to get a pending selector it will throw and then restart when the dependency resolves or errors.

Perfect! That's what I hoped would happen. Hadn't clicking that the selector not retrying was related to the 'non-subscribing' issue. Sorry @drarmstr !

**Comment by drarmstr** [comment]
Duplicate of #103

**Comment by acutmore** [comment]
That aside there is still a small syntax typo, with a missing bracket.
